### PR TITLE
HTCONDOR-2004: Add checks for upgrading to V23

### PR DESCRIPTION
**Pull request by ColeBollig**
-Added check for systems using fragmented desktop policy macros -Added warning for deprecation of old job router config syntax
 slated for removal in V23 feature series lifetime

<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Comment by ColeBollig** [comment]
I already ran this through python black

File: src/condor_scripts/condor_upgrade_check [hunk 1]
```diff
@@ -550,6 +550,179 @@ def warn_user_records():
...
+    return 0
+
+
+def check_desktop_policy():
+    """Check for use of desktop policy configuration knobs that became hidden in V10.4.0"""
+    htc_version = htcondor.param["CONDOR_VERSION"].split(".")
+    if int(htc_version[0]) < 10 or (
+        int(htc_version[0]) == 10 and int(htc_version[1]) < 4
+    ):
+        # Config items formally avaible 100% of the time
```
Comment (jasoncpatton) [inline src/condor_scripts/condor_upgrade_check:600]: I think you mean this?
```suggestion
        # Config items formerly available 100% of the time
```

File: src/condor_scripts/condor_upgrade_check [hunk 2]
```diff
@@ -550,6 +550,179 @@ def warn_user_records():
...
+                            pos = check.find("cpubusy", pos + 1)
+                        if false_positive:
+                            continue
+                    not_in_other_def = True
+                    for other in attrs:
+                        if other.lower() == parts[0].strip():
+                            not_in_other_def = False
+                            break
+                    if not_in_other_def:
+                        found_attrs.append(attr)
```
Comment (jasoncpatton) [inline src/condor_scripts/condor_upgrade_check:644]: Using a set for `found_attrs` seems more appropriate as we only care to know if they're using a forbidden `attr` at least once and don't want to print out the same `attr` mutliple times later.
```suggestion
        found_attrs = set()
        has_cpu_busy_time = []

        p = subprocess.run(["condor_config_val", "-dump"], stdout=subprocess.PIPE)
        cmd_output = p.stdout.rstrip().decode()
        has_problem = False
        for line in cmd_output.split("\n"):
            line = line.strip()
            check = line.lower()
            parts = check.split("=", 1)
            if len(parts) < 2:
                continue
            for attr in attrs:
                if attr.lower() in parts[1]:
                    if attr == "CpuBusy":
                        pos = check.find("cpubusy")
                        false_positive = True
                        while pos != -1:
                            if check.find("cpubusytime", pos) != pos:
                                false_positive = False
                                break
                            pos = check.find("cpubusy", pos + 1)
                        if false_positive:
                            continue
                    not_in_other_def = True
                    for other in attrs:
                        if other.lower() == parts[0].strip():
                            not_in_other_def = False
                            break
                    if not_in_other_def:
                        found_attrs.add(attr)
```

Most other methods used later (`found_attrs.clear()`, `found_attrs.remove()`) work the same for sets, except for `found_attrs.sort()`.

File: src/condor_scripts/condor_upgrade_check [hunk 3]
```diff
@@ -550,6 +550,179 @@ def warn_user_records():
...
+        for attr in custom_set:
+            found_attrs.remove(attr)
+
+        if has_problem:
+            format_print(
+                "Error checking for desktop policy configuration macros:", newline=True
+            )
+            need_newline = False
+            if len(found_attrs) > 0:
+                found_attrs.sort()
```
Comment (jasoncpatton) [inline src/condor_scripts/condor_upgrade_check:689]: `sorted()` will convert a set to a list and sort it.
```suggestion
                found_attrs = sorted(found_attrs)
```
Consider batching this with the previous suggestion.

File: src/condor_scripts/condor_upgrade_check [hunk 4]
```diff
@@ -550,6 +550,179 @@ def warn_user_records():
...
+        return 1
+    return 0
+
+
+def check_desktop_policy():
+    """Check for use of desktop policy configuration knobs that became hidden in V10.4.0"""
+    htc_version = htcondor.param["CONDOR_VERSION"].split(".")
+    if int(htc_version[0]) < 10 or (
+        int(htc_version[0]) == 10 and int(htc_version[1]) < 4
+    ):
```
Comment (jasoncpatton) [inline src/condor_scripts/condor_upgrade_check:599]:
```suggestion
    htc_version = tuple(int(x) for x in htcondor.param["CONDOR_VERSION"].split("."))
    if htc_version < (10, 4, 0):
```

Use tuples for numeric version comparison:

```python
>>> htc_version = tuple(int(x) for x in htcondor.param["CONDOR_VERSION"].split("."))
>>> htc_version
(10, 8, 0)
>>> htc_version < (10, 0, 0)
False
>>> htc_version < (10, 10, 0)
True
>>> htc_version < (11, 0, 0)
True```

**Review by jasoncpatton — CHANGES_REQUESTED** [review]
I've got a few minor suggested changes, otherwise looks good.

**Pull request merged**